overlord/snapstate/target.go

```diff
@@ -0,0 +1,536 @@
+		// caller requested that we ignore validation sets, nothing to do
+		action.Flags = store.SnapActionIgnoreValidation
+	case len(t.RevOpts.ValidationSets) > 0:
+		// caller provided some validation sets, nothing to do but send them
+		// to the store
+		action.ValidationSets = t.RevOpts.ValidationSets
+	default:
+		vsets, err := enforcedSets()
+		if err != nil {
+			return nil, err
+		}
+
+		// if the caller didn't provide any validation sets, make sure that
+		// the snap is allowed by all of the enforced validation sets
+		invalidSets, err := vsets.CheckPresenceInvalid(naming.Snap(t.Name))
+		if err != nil {
+			if _, ok := err.(*snapasserts.PresenceConstraintError); !ok {
+				return nil, err
+			} // else presence is optional or required, carry on
+		}
+
+		if len(invalidSets) > 0 {
+			return nil, fmt.Errorf(
+				"cannot install snap %q due to enforcing rules of validation set %s",
+				t.Name, snapasserts.ValidationSetKeySlice(invalidSets).CommaSeparated(),
+			)
+		}
+
+		requiredSets, requiredRev, err := vsets.CheckPresenceRequired(naming.Snap(t.Name))
+		if err != nil {
+			return nil, err
+		}
+
+		// make sure that the caller-requested revision matches the revision
+		// required by the enforced validation sets
+		if requiredRev.Set() && t.RevOpts.Revision.Set() && requiredRev != t.RevOpts.Revision {
+			return nil, fmt.Errorf(
+				"cannot install snap %q at requested revision %s without --ignore-validation, revision %s required by validation sets: %s",
+				t.Name, t.RevOpts.Revision, requiredRev, snapasserts.ValidationSetKeySlice(requiredSets).CommaSeparated(),
+			)
+		}
+
+		// TODO: handle validation sets and components here
+
+		action.ValidationSets = requiredSets
+
+		if requiredRev.Set() {
+			// make sure that we use the revision required by the enforced
+			// validation sets
+			action.Revision = requiredRev
+
+			// we ignore the cohort if a validation set requires that the
+			// snap is pinned to a specific revision
+			action.CohortKey = ""
+		}
+	}
+
+	// clear out the channel if we're requesting a specific revision, which
+	// could be because the user requested a specific revision or because a
+	// validation set requires it
+	if action.Revision.Set() {
+		action.Channel = ""
+	}
+
+	return action, nil
+}
+
+func (s *StoreTarget) validateAndPrune(snaps map[string]*SnapState) error {
+	for name, t := range s.snaps {
+		if err := snap.ValidateInstanceName(name); err != nil {
+			return fmt.Errorf("invalid instance name: %v", err)
+		}
+
+		if err := validateRevisionOpts(t.RevOpts); err != nil {
+			return fmt.Errorf("invalid revision options for snap %q: %w", name, err)
+		}
+
+		snapst, ok := snaps[name]
+		if ok && snapst.IsInstalled() {
+			if !t.SkipIfPresent {
+				return &snap.AlreadyInstalledError{Snap: name}
+			}
+			delete(s.snaps, name)
+		}
+	}
+	return nil
+}
+
+func InstallOne(ctx context.Context, st *state.State, target Target, opts SnapstateOptions) (*snap.Info, *state.TaskSet, error) {
```

Comment:
Comparing `InstallOne` with `InstallTarget` by their names, it's not clear which one does what. Especially since `Target` can be one or many snaps

Reply:
Hrm, you're right. I'll think about names. In an ideal world, `InstallTarget` will become `Install` one day, once the old API stops being used.